docs/source/install.rst

```diff
@@ -0,0 +1,316 @@
+is installed with a specific set of external packages. To install PETSc you need to
+do the following steps:
+
+#. Clone the PETSc repository::
+
+   $ git clone https://github.com/firedrakeproject/petsc.git
+   $ cd petsc
+
+#. Run PETSc ``configure``, passing in the flags generated by ``firedrake-configure``::
+
+   $ ./configure $(python3 ../firedrake-configure --show-petsc-configure-options)
+
+#. Compile PETSc::
+
+   $ make
+
+If you are using one of the
+:ref:`officially supported distributions<supported_systems>` then these configure
+options will include paths to system packages so PETSc can correctly find and
+link against them. If you are not then you should pass the ``--no-package-manager``
+flag to obtain a set of configure options where ``firedrake-configure``
+pessimistically assumes that no external packages are available, and hence need
+to be downloaded and compiled from source::
+
+   $ ./configure $(python3 ../firedrake-configure --no-package-manager --show-petsc-configure-options)
+
+For the ``default`` ARCH, running ``firedrake-configure`` with
+``--no-package-manager`` will produce the flags:
+
+.. literalinclude:: petsc_configure_options.txt
+   :language: text
+
+
+Customising the PETSc installation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Since ``firedrake-configure`` only outputs a string of options it is straightforward
+to make changes to the options passed to PETSc ``configure``. You can either:
+
+* Append additional options when ``configure`` is invoked. For example, to compile
+  PETSc with 64 bit indices you should run::
+
+   $ ./configure $(python3 ../firedrake-configure --show-petsc-configure-options) --with-64-bit-indices
+
+* Write the output of ``firedrake-configure`` to a file than can be modified::
+
+   $ python3 ../firedrake-configure --show-petsc-configure-options > my_configure_options.txt
+   <edit my_configure_options.txt>
+   $ ./configure $(cat my_configure_options.txt)
+
+
+.. _install_firedrake:
+
+Installing Firedrake
+--------------------
+
+Now that the right system packages are installed and PETSc is built we can now
+install Firedrake. To do this perform the following steps:
+
+#. Create a `virtual environment <https://docs.python.org/3/tutorial/venv.html>`_::
+
+      $ python3 -m venv venv-firedrake
+      $ . venv-firedrake/bin/activate
+
+   This is optional but strongly recommended to avoid polluting your system Python
+   environment.
+
+#. Set any necessary environment variables. This can be achieved using
+   ``firedrake-configure``::
+
+     $ export $(python3 firedrake-configure --show-env)
+
+   At a minimum this will set the following variables:
+
+   .. code-block:: text
+
+      CC=/path/to/mpicc CXX=/path/to/mpicxx PETSC_DIR=/path/to/petsc PETSC_ARCH=arch-firedrake-{default,complex} HDF5_MPI=ON
+
+#. Install Firedrake::
+
+     $ pip install --no-binary h5py 'git+https://github.com/firedrakeproject/firedrake.git[test]'
```

Comment:
A note to say that, whereas with `firedrake-install` the default was to have an editable install, this will not give you and editable install and you will have to specify if that is what you want.
Possibly obvious, but the end result for someone following the installation instructions to the letter has changed.

Reply:
Yeah I should mention something here

Reply:
Also that, even if firedrake is installed in editable mode, all the other packages will not be, e.g. pyadjoint, ufl, etc.

I'm assuming that if I want a "full" editable install then I should do `pip install -e ufl pyadjoint ...` (with the properly addressed package location) before `pip install firedrake`?

Reply:
I haven't mentioned getting editable version of subpackages. I guess I should...

The order of install doesn't actually matter. You can install Firedrake first with no issues

Reply:
But wouldn't pip installing firedrake mean that pip would install all of Firedrake's dependencies? So you would then have to uninstall them, then do the editable install. Doable, but an extra step needlessly

Reply:
> I have added a section titled "Developer install" that explains how to do that.

Great, thanks.

Reply:
> But wouldn't pip installing firedrake mean that pip would install all of Firedrake's dependencies? So you would then have to uninstall them, then do the editable install. Doable, but an extra step needlessly

I believe that you have to do it that way around. Otherwise when you install Firedrake it will try and replace the already installed package.

Reply:
Oh, that's not what I'd expect. Is that a pip thing or a firedrake thing? Wouldn't pip look for the dependencies, see they are already satisfied, and move on?

Reply:
But the versions aren't satisfied. pip can't reason about non-versioned packages so it always assumes that they are out-of-date and reinstalls them

Reply:
This is mentioned in the "Developer install" section now.

Reply:
> But the versions aren't satisfied. pip can't reason about non-versioned packages so it always assumes that they are out-of-date and reinstalls them

Gotcha.